base/meta.jl

```diff
@@ -46,4 +89,151 @@ function show_sexpr(io::IO, ex::Expr, indent::Int)
+    pat == ex || nomatch(pat, ex)
+    return env
+end
+
+match_inner(pat::QuoteNode, ex::QuoteNode, env) =
+match(pat.value, ex.value, env)
+
+isslurp(s) = false
+isslurp(s::Symbol) = s == :__ || Base.ismatch(r"[^_]__$", string(s))
+
+function slurprange(pat)
+    slurps = length(filter(isslurp, pat))
+    slurps == 0 && return 0,0
+    slurps > 1 && error("Pattern may only contain one slurp.")
+
+    left, right = 1, 1
+    while !isslurp(pat[left]) left += 1 end
+    while !isslurp(pat[end+1-right]) right += 1 end
+    return left, right
+end
+
+inrange(i, range, len) =
+    range ≠ (0,0) && i ≥ range[1] && i ≤ len+1-range[2]
+
+function match_inner(pat::Expr, ex::Expr, env)
+    match(pat.head, ex.head, env)
+    pat, ex = rmlines(pat), rmlines(ex)
+    sr = slurprange(pat.args)
+    slurp = Any[]
+    i = 1
+    for p in pat.args
+        i > length(ex.args) &&
+            (isslurp(p) ? (env[bname(p)] = slurp) : nomatch(pat, ex))
+
+        while inrange(i, sr, length(ex.args))
+            push!(slurp, ex.args[i])
+            i += 1
+        end
+
+        if isslurp(p)
+            p ≠ :__ && (nv[bname(p)] = slurp)
+        else
+            match(p, ex.args[i], env)
+            i += 1
+        end
+    end
+    i == length(ex.args)+1 || nomatch(pat, ex)
+    return env
+end
+
+blockunify(a, b) =
+    isexpr(a, :block) && !isexpr(b, :block) ? (a, Expr(:block, b)) :
+    !isexpr(a, :block) && isexpr(b, :block) ? (Expr(:block, a), b) :
+    (a, b)
+
+function match(pat, ex, env)
+    pat, ex = unblock(pat), unblock(ex)
+    pat == :_ && return env
+    isbinding(pat) && return (env[bname(pat)] = ex; env)
+    pat, ex = blockunify(pat, ex)
+    return match_inner(pat, ex, env)
+end
+
+match(pat, ex) = match(pat, ex, Dict())
+
+function ismatch(pat, ex)
+    try
+        match(pat, ex)
+        return true
+    catch e
+        isa(e, MatchError) ? (return false) : rethrow()
+    end
+end
+
+function trymatch(pat, ex)
+    try
+        match(pat, ex)
+    catch e
+        isa(e, MatchError) ? (return) : rethrow()
+    end
+end
+
+# @match macro
+
+allbindings(pat, bs) =
+    isbinding(pat) || (isslurp(pat) && pat ≠ :__) ? push!(bs, bname(pat)) :
+    isexpr(pat, :$) ? bs :
+    isa(pat, Expr) ? map(pat -> allbindings(pat, bs), [pat.head, pat.args...]) :
+    bs
+
+allbindings(pat) = (bs = Any[]; allbindings(pat, bs); bs)
+
+function bindinglet(bs, body)
+    ex = :(let $(esc(:env)) = env
+               $body
+           end)
+    for b in bs
+        push!(ex.args, :($(esc(b)) = env[$(Expr(:quote, b))]))
+    end
+    return ex
+end
+
+function makeclause(line, els = nothing)
+    env = trymatch(:(pat_ -> yes_), line)
+    env == nothing && error("Invalid match clause $line")
+    pat, yes = env[:pat], env[:yes]
+    quote
+        env = trymatch($(Expr(:quote, pat)), ex)
+        if env != nothing
+            $(bindinglet(allbindings(pat), esc(yes)))
+        else
+            $els
+        end
+    end
+end
+
+macro match (ex, lines)
```

Comment:
Which is a shame, because I likes this style of writing macros (it matches the call better). I guess I'm not going to convince anyone else of that, though.